### PR TITLE
 Improved error handling, especially for celery tasks

### PR DESCRIPTION
**Pull request by mbertrand**
#### What are the relevant tickets?
Related to https://github.com/mitodl/odl-video-service/issues/24 and https://github.com/mitodl/odl-video-service/issues/58

#### What's this PR do?
- New statuses for tasks to better track progress (Uploading, Transcoding, Upload Failed, Transcode Failed).
- Overrides default Transcoder.create_job class to record more detailed results info in EncodeJob objects
- Uses try-catch blocks to update statuses of videos and tasks when failures occur in celery tasks.
- Fixes progress bar on upload page.

#### How should this be manually tested?
- Upload a video file that's about 50 MB or larger.  Progress bar should begin filling in shortly.  Look at network requests and responses: responses should show amount of bytes transferred. When upload is complete, requests should stop being sent and the last response should show a status of 'Transcoding'.  Go to the video detail page, the video 'Encoding Status' should also be 'Transcoding'.  When transcoding is finished, the status should be shown as 'Complete'
- Choose a file to upload from dropbox, then rename the dropbox file after it appears on the upload list but before the progress bar starts filling in.  Eventually, after 15 minutes or so, the upload should fail after AWS gives up trying. Go to the video detail page and the 'Encoding Status' should be 'Upload failed'
- Rename a non-video dropbox file as "myvideo.mp4" and upload it.  Upload should complete successfully but transcoding should eventually fail, and the 'Encoding status' on the video detail page should be 'Transcoding failed'
- Go to ../admin/dj_elastictranscoder/encodejob/ and you should see EncodeJob objects for all videos whose transcoding tasks passed or failed, along with messages that contain the job results details.

### CODE REVIEW

File: cloudsync/tasks.py [hunk 1]
```diff
@@ -23,58 +26,92 @@
...
+        response = requests.get(video.source_url, stream=True, timeout=60)
+        response.raise_for_status()
+        _, content_type, content_length = parse_content_metadata(response)
+
+        s3 = boto3.resource('s3')
+        bucket_name = settings.VIDEO_S3_BUCKET
+        bucket = s3.Bucket(bucket_name)
+        total_bytes_uploaded = 0
+
+        def callback(bytes_uploaded):
```
Comment (giocalitri) [inline cloudsync/tasks.py]: Should we move this function definition outside the scope of the other function? It does not make any sense to have it here.

File: cloudsync/tasks.py [hunk 2]
```diff
@@ -23,58 +26,92 @@
...
+            Key=video.s3_key(),
+            ExtraArgs={"ContentType": content_type},
+            Callback=callback,
+            Config=config
+        )
+    except Exception as e:
+        log.exception('Upload failed for video %d', video_id)
+        update_video_status(video, VideoStatus.UPLOAD_FAILED)
+        self.update_state(task_id=task_id, state=states.FAILURE)
+        raise e
```
Comment (giocalitri) [inline cloudsync/tasks.py]: I do not think that the approach of wrapping the entire code in a try/except is a good one.
I think you need to wrap just the single call you are afraid to fail.
In this way it would be complicated to find where the code actually failed.

File: cloudsync/tasks.py [hunk 3]
```diff
@@ -23,58 +26,92 @@
...
+@shared_task(bind=True, base=VideoTask)
 def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
     """
     Given an S3 object key, transcode that object using a video pipeline, overwrite the original when done.
 
     Args:
         video_id(int): The video primary key
     """
+    video = Video.objects.get(id=video_id)
+    update_video_status(video, VideoStatus.TRANSCODING)
```
Comment (giocalitri) [inline cloudsync/tasks.py]: cannot `update_video_status` simple be a method of the Video model?

File: cloudsync/tasks.py [hunk 4]
```diff
@@ -95,14 +132,36 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
-    transcoder.encode(video_input, outputs, Playlists=playlists)
-    transcoder.create_job_for_object(video)
+    try:
+        transcoder.encode(video_input, outputs, Playlists=playlists)
+    except ClientError as e:
+        log.exception('Transcode job creation failed for video %s: %s', video_id, e.response)
+        update_video_status(video, VideoStatus.TRANSCODE_FAILED)
+        self.update_state(task_id=task_id, state=states.FAILURE)
+        transcoder.message = e.response
+        raise e
```
Comment (giocalitri) [inline cloudsync/tasks.py]: if you raise anyway, why do you `'log.exception`?

File: cloudsync/tasks.py [hunk 5]
```diff
@@ -95,14 +132,36 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
+    finally:
+        job = transcoder.create_job_for_object(video)
```
Comment (giocalitri) [inline cloudsync/tasks.py]: what does this do? A retry?

File: cloudsync/tasks.py [hunk 6]
```diff
@@ -95,14 +132,36 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
+
+    # Wait for the transcoding job to complete
+    waiter = transcoder.client.get_waiter('job_complete')
+    try:
+        waiter.wait(Id=transcoder.message['Job']['Id'])
```
Comment (giocalitri) [inline cloudsync/tasks.py]: is this locking on the request?

File: ui/api.py [hunk 7]
```diff
@@ -53,19 +55,42 @@ def process_transcode_results(video, job):
...
+            if not encode_job:
+                encode_job = video.encode_jobs.latest("created_at")
+            et_job = get_et_job(encode_job.id)
+            if et_job['Status'] == VideoStatus.COMPLETE:
+                process_transcode_results(video, et_job)
+                encode_job.message = et_job
+                encode_job.save()
+                update_video_status(video, VideoStatus.COMPLETE)
+            elif et_job['Status'] == VideoStatus.ERROR:
+                update_video_status(video, VideoStatus.TRANSCODE_FAILED)
```
Comment (giocalitri) [inline ui/api.py]: I do not think that wrapping all this code in a try/except is a good idea.
the `EncodeJob.DoesNotExist` seems to be relate to just one call and so does `ClientError`

File: ui/api.py [hunk 8]
```diff
@@ -53,19 +55,42 @@ def process_transcode_results(video, job):
...
+
+
+def update_video_status(video, status):
+    """
+    Assign and save the status of a Video
+    :param video: The video to update
+    :param status: The status to assign the video
+    """
+    video.status = status
+    video.save()
```
Comment (giocalitri) [inline ui/api.py]: yes, this should be a method of the video model

File: ui/api_test.py [hunk 9]
```diff
@@ -19,10 +20,24 @@ def test_video_job_status(mocker, video, encodejob):  # pylint: disable=unused-a
...
+    """
+    Verify that refresh_status does not raise ClientError
+    """
+    video.status = VideoStatus.TRANSCODING
+    mocker.patch('ui.util.boto3', MockBoto)
+    error = Exception("unexpected exception")
+    mocker.patch('ui.util.get_transcoder_client',
+                 return_value=MockClientET(error=error))
+    with pytest.raises(Exception):
+        refresh_status(video)
```
Comment (giocalitri) [inline ui/api_test.py]: Question: are there already tests for all the use cases of the API functions?

File: ui/statuses.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+
+class VideoStatus:
+    """Simple class for possible video statuses"""
+    UPLOADING = 'Uploading'
+    UPLOAD_FAILED = 'Upload failed'
+    TRANSCODING = 'Transcoding'
+    TRANSCODE_FAILED = 'Transcode failed'
+    COMPLETE = 'Complete'
+    ERROR = 'Error'
```
Comment (giocalitri) [inline ui/statuses.py]: What about rename the module from `statuses.py` to `constants.py`

File: cloudsync/tasks.py [hunk 11]
```diff
@@ -23,58 +26,92 @@
...
-
-    bucket.upload_fileobj(
-        Fileobj=response.raw,
-        Key=s3_key,
-        ExtraArgs={"ContentType": content_type},
-        Callback=callback,
-        Config=config
-    )
+        if self.request.chain:
+            return self.request.chain[0]['options']['task_id']
```
Comment (mbertrand) [inline cloudsync/tasks.py]: self.request.chain[0]['options']['task_id'] is what correctly returns the id of the task during upload and allows the progress bar to accurately reflect the percentage of the file uploaded from dropbox to s3.  Using request.id doesn't work when the upload and transcode tasks are chained.

File: cloudsync/tasks.py (hunk 1 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: It is an inner function due to scope issues, it needs access to `video`, `content_length`, and `total_bytes_uploaded`.

File: cloudsync/tasks.py (hunk 2 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: ok, I will put try/except just around  `bucket.upload_fileobj` and `response.raise_for_status()`.

File: cloudsync/tasks.py (hunk 5 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: This creates an [EncodeJob](https://github.com/mitodl/django-elastic-transcoder/blob/master/dj_elastictranscoder/models.py#L7) object to represent the transcode job and its status/results.

File: cloudsync/tasks.py (hunk 6 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: It makes `read_job` requests to AWS every 30 seconds to see if the transcode job has finished.  So it keeps the celery task running until then.  The more I think about it the less I like this approach.  I would prefer to have a separate periodic celery task that would run every X seconds/minutes to check on the job status instead.

File: cloudsync/tasks.py (hunk 3 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: sure

File: cloudsync/tasks.py (hunk 4 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: I'll get rid of the log statements

File: ui/api_test.py (hunk 9 above)
Comment (mbertrand) [inline ui/api_test.py]: No: https://github.com/mitodl/odl-video-service/issues/17

File: ui/statuses.py (hunk 10 above)
Comment (mbertrand) [inline ui/statuses.py]: sure

File: ui/api.py (hunk 7 above)
Comment (mbertrand) [inline ui/api.py]: ClientError could happen in either `get_et_job` or `process_transcode_results` method calls.  I'll make a separate try/catch block for `EncodeJob.DoesNotExist`

File: cloudsync/tasks.py (hunk 4 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Actually the log statements provide info on the video id for which the exception occurred, which may be useful for troubleshooting. Should I get rid of them anyway?

File: cloudsync/tasks.py (hunk 11 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: I understand that `request.id` doesn't work when the upload and transcode tasks are chained, but again, the documentation seems to suggest that the current task is not `0` in the chain

File: cloudsync/tasks.py [hunk 12]
```diff
@@ -23,58 +26,95 @@
...
-    response.raise_for_status()
+
+    video = Video.objects.get(id=video_id)
+    video.update_status(VideoStatus.UPLOADING)
+
+    task_id = self.get_task_id()
+    try:
+        response = requests.get(video.source_url, stream=True, timeout=60)
+        response.raise_for_status()
+    except Exception as e:
```
Comment (giocalitri) [inline cloudsync/tasks.py]: nit: please do not use single letters as variable names

Comment (giocalitri) [inline cloudsync/tasks.py]: this comment also for the rest of the PR

File: cloudsync/tasks.py [hunk 13]
```diff
@@ -95,14 +135,37 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
         settings.AWS_ACCESS_KEY_ID,
         settings.AWS_SECRET_ACCESS_KEY
     )
-    transcoder.encode(video_input, outputs, Playlists=playlists)
-    transcoder.create_job_for_object(video)
+    try:
+        if not settings.ET_PRESET_IDS:
+            raise ValueError("At least one transcode preset required in settings")
+        transcoder.encode(video_input, outputs, Playlists=playlists)
+    except (ValueError, ClientError) as e:
```
Comment (giocalitri) [inline cloudsync/tasks.py]: nit: if you are using tuples, please end the "list" with a comma to make clear that it is a tuple
`(ValueError, ClientError,)`

File: cloudsync/tasks.py [hunk 14]
```diff
@@ -95,14 +135,37 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
+        log.exception('Transcode job creation failed for video %s: %s', video_id, e.response)
```
Comment (giocalitri) [inline cloudsync/tasks.py]: doesn't this duplicate the stack trace if later you raise anyway?

File: cloudsync/tasks.py (hunk 5 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: and why it is a method of the transcoder?

File: cloudsync/tasks.py (hunk 6 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: what you described is locking the celery worker.
we probably should skip the execution and enqueue a new task with a delay: something like this
http://docs.celeryproject.org/en/latest/userguide/calling.html#eta-and-countdown

File: cloudsync/tasks.py (hunk 11 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: `"1a859e5a-8f71-4e01-9349-5ebc6dc66631"` is the parent task?

File: cloudsync/tasks.py (hunk 6 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Related to #58 

If the `waiter.wait` method is put in a new delayed task, that task will then be locked until `waiter` finally gets a response, right?  What do you think of a periodic task that checks the status of all videos that have not yet completed or failed?  Or we could make use of AWS SNS with django-elastic-transcoder's endpoint API, though that makes things harder to test locally.

File: cloudsync/tasks.py (hunk 1 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: isn't a way to pass arguments to the callback?

File: cloudsync/tasks.py (hunk 11 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: I think so, or it effectively seems to function as one anyway.  That is the task id returned by the uploader view response.  There is a `parent_id` attribute in the task.request but it is `None`, and also an `id`, `root_id` and `correlation_id` but none of them are equal to `1a859e5a-8f71-4e01-9349-5ebc6dc66631`

File: cloudsync/tasks.py (hunk 4 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: maybe change it to a `log.error` if you want to keep the info

File: ui/api.py (hunk 7 above)
Comment (giocalitri) [inline ui/api.py]: ok

File: cloudsync/tasks.py (hunk 1 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: The only argument that seems to be passed by `Bucket.upload_fileobj` is the number of bytes transferred: http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Bucket.upload_fileobj

File: cloudsync/tasks.py (hunk 11 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: ok then
but please wrap the call with a try/except so you can log what went wrong in case the structure changes.
Also you should write tests for this

File: cloudsync/tasks.py (hunk 14 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Right, I'll take out the stack trace from the log message

File: cloudsync/tasks.py [hunk 15]
```diff
@@ -95,14 +135,37 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
+        video.update_status(VideoStatus.TRANSCODE_FAILED)
+        self.update_state(task_id=task_id, state=states.FAILURE)
+        transcoder.message = e.response
```
Comment (noisecapella) [inline cloudsync/tasks.py]: Sorry for the drive by review, but `e.response` won't exist for `ValueError`

File: cloudsync/tasks.py [hunk 16]
```diff
@@ -95,14 +135,37 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
+        raise e
```
Comment (noisecapella) [inline cloudsync/tasks.py]: You can just do `raise` here

File: cloudsync/tasks.py (hunk 15 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Thanks, will fix

File: cloudsync/tasks.py (hunk 6 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: @giocalitri what do you think of something like this, running every x minutes:

```
@shared_task(bind=True)
def update_video_statuses(self):
    transcoding_videos = Video.objects.filter(status=VideoStatus.TRANSCODING)
    for video in transcoding_videos:
        try:
            encode_job = video.encode_jobs.latest("created_at")
        except EncodeJob.DoesNotExist as e:
            # This is OK if video is still being uploaded
            if video.status != VideoStatus.UPLOADING:
                log.exception("No EncodeJob object exists for video id %d", video.id)
                raise
        try:
            et_status = get_et_job(encode_job.id)
            encode_job.message = et_status
            encode_job.save()
            if et_status['Status'] == VideoStatus.COMPLETE:
                process_transcode_results(video, et_status)
                video.update_status(VideoStatus.COMPLETE)
            elif et_status['Status'] == VideoStatus.ERROR:
                log.exception('Encoding failed for video %d', video.id)
                video.update(VideoStatus.TRANSCODE_FAILED)
        except ClientError as e:
            log.exception("AWS error when refreshing job status for video %d", video.id)
            raise
```

Comment (mbertrand) [inline cloudsync/tasks.py]: Maybe this should be moved to a separate PR?

Comment (mbertrand) [inline cloudsync/tasks.py]: Decided to go with option 3 from #58 based on our discussion this AM

File: cloudsync/tasks.py [hunk 17]
```diff
@@ -23,58 +27,100 @@
...
+        Returns:
+            The request id
+        """
+        if self.request.chain:
+            try:
+                return self.request.chain[0]['options']['task_id']
+            except (IndexError, KeyError,) as exc:
+                # Log the error and continue, using self.request.id instead
+                # The worst that will happen is that progress bar won't work.
+                log.error("Could not find task_id in chain: %s", str(exc))
```
Comment (giocalitri) [inline cloudsync/tasks.py]: this will make the function return `self.request.id` that I do not think it is what should do

File: cloudsync/tasks.py [hunk 18]
```diff
@@ -23,58 +27,100 @@
...
     Stream the contents of the given URL to Amazon S3
     """
-    if not url:
+
+    if not video_id:
         return False
-    response = requests.get(url, stream=True)
-    response.raise_for_status()
+
+    video = Video.objects.get(id=video_id)
```
Comment (giocalitri) [inline cloudsync/tasks.py]: this could theoretically raise: if you want to avoid, use try/except or `.filter` and `.first`

File: cloudsync/tasks.py (hunk 17 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Make it return `None` instead?  This is only required for making the progress bar work.

File: cloudsync/tasks.py [hunk 19]
```diff
@@ -23,58 +27,100 @@
...
 
     def callback(bytes_uploaded):
         """
         Callback function after upload
         """
+        nonlocal total_bytes_uploaded
+        total_bytes_uploaded = total_bytes_uploaded + bytes_uploaded
         data = {
-            "uploaded": bytes_uploaded,
+            "uploaded": total_bytes_uploaded + bytes_uploaded,
```
Comment (giocalitri) [inline cloudsync/tasks.py]: why do you sum again here?

File: cloudsync/tasks.py [hunk 20]
```diff
@@ -23,58 +27,100 @@
...
-        Callback=callback,
-        Config=config
-    )
-
-
-@shared_task(bind=True)
+    try:
+        bucket.upload_fileobj(
+            Fileobj=response.raw,
+            Key=video.s3_key(),
```
Comment (giocalitri) [inline cloudsync/tasks.py]: we should probably make `Video.s3_key()` a property method

File: cloudsync/tasks.py (hunk 19 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: To get the number of bytes transferred since the upload began, so it can be compared to the file size (total), and percentage progress calculated.

File: cloudsync/tasks.py [hunk 21]
```diff
@@ -95,14 +141,42 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
         settings.ET_PIPELINE_ID,
         settings.AWS_REGION,
         settings.AWS_ACCESS_KEY_ID,
         settings.AWS_SECRET_ACCESS_KEY
     )
-    transcoder.encode(video_input, outputs, Playlists=playlists)
-    transcoder.create_job_for_object(video)
+    try:
+        if not settings.ET_PRESET_IDS:
+            raise ValueError("At least one transcode preset required in settings")
```
Comment (giocalitri) [inline cloudsync/tasks.py]: why this is not in its own try/except?

Comment (giocalitri) [inline cloudsync/tasks.py]: to reuse the same code in the except?

File: cloudsync/tasks.py [hunk 22]
```diff
@@ -95,14 +141,42 @@ def transcode_from_s3(self, video_id):  # pylint: disable=unused-argument
...
+    """
+    Check on statuses of all transcoding videos and update their status if appropriate
+    """
+    transcoding_videos = Video.objects.filter(status=VideoStatus.TRANSCODING)
+    for video in transcoding_videos:
+        try:
+            refresh_status(video)
+        except EncodeJob.DoesNotExist:
+            # Log the exception but don't raise it so other videos can be checked.
+            log.exception("No EncodeJob object exists for video id %d", video.id)
```
Comment (giocalitri) [inline cloudsync/tasks.py]: don't we need to do anything else in this case?

File: cloudsync/tasks.py (hunk 20 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Originally it was, but [based on feedback from an earlier code review](https://github.com/mitodl/odl-video-service/pull/11#discussion_r124569764), it was changed to a function instead.  But I can change it back to a property if you prefer.

File: cloudsync/tasks.py (hunk 17 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: not sure how it is used: if None does not crash anything else, go for it

File: cloudsync/tasks.py (hunk 22 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: I don't want to raise an exception here because it might block status checking for other videos.  Sentry.io should alert us of the error in the log, right?

File: cloudsync/tasks.py (hunk 20 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: you are right, sorry, but maybe a name like `get_s3_key` would be better for a function

File: cloudsync/tasks.py (hunk 21 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: Yes, to avoid having much of the same code in the except parts (`video.update()`, `self.update_state()`)

File: cloudsync/tasks.py (hunk 20 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: sure, I'll rename it

File: cloudsync/tasks.py (hunk 22 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: I could change the video status to 'Failed' here if there is no EncodeJob

File: cloudsync/tasks.py (hunk 21 above)
Comment (mbertrand) [inline cloudsync/tasks.py]: I think I'll move this check directly to the `settings.py` file, so it will raise an error immediately on startup if missing.

Comment (noisecapella) [inline cloudsync/tasks.py]: I'm not sure you can do that, it interferes with `docker-compose build` because it runs a management command `collectstatic` which will import the settings

Comment (mbertrand) [inline cloudsync/tasks.py]: Since there are defaults for `ET_PRESET_IDS` in settings now, this shouldn't ever happen anyway. So it should probably just be removed.

File: cloudsync/tasks.py [hunk 23]
```diff
@@ -23,58 +27,103 @@
...
+            The request id
+        """
+        if self.request.chain:
+            try:
+                return self.request.chain[0]['options']['task_id']
+            except (IndexError, KeyError,) as exc:
+                # Log the error and continue, using self.request.id instead
+                # The worst that will happen is that progress bar won't work.
+                log.error("Could not find task_id in chain: %s", str(exc))
+                return None
```
Comment (giocalitri) [inline cloudsync/tasks.py]: nit: just `return` (None is implicit)

File: cloudsync/tasks.py (hunk 22 above)
Comment (giocalitri) [inline cloudsync/tasks.py]: ok

File: cloudsync/util.py [hunk 24]
```diff
@@ -0,0 +1,34 @@
+"""Utility classes/methods for cloudsync"""
```
Comment (giocalitri) [inline cloudsync/util.py]: please name the module to `utils`

File: odl_video/settings.py [hunk 25]
```diff
@@ -440,11 +445,15 @@ def get_all_config_keys():
     '1351620000001-200010,1351620000001-200020,1351620000001-200050'
 ).split(',')
 
+if ET_PRESET_IDS == ['']:  # This may happen if `ET_PRESET_IDS=` is in .env file.
+    raise ValueError('ET_PRESET_IDS cannot be blank, please check your settings & environment')
```
Comment (giocalitri) [inline odl_video/settings.py]: probably [ImproperlyConfigured](https://docs.djangoproject.com/en/1.11/ref/exceptions/#improperlyconfigured) is better here

File: ui/api.py [hunk 26]
```diff
@@ -53,19 +53,22 @@ def process_transcode_results(video, job):
...
-            log.error("No encoding job exists for video id %d", video.id)
+    if video.status == VideoStatus.TRANSCODING:
+        if not encode_job:
+            encode_job = video.encode_jobs.latest("created_at")
+        et_job = get_et_job(encode_job.id)
+        if et_job['Status'] == VideoStatus.COMPLETE:
+            process_transcode_results(video, et_job)
+            video.update_status(VideoStatus.COMPLETE)
+        elif et_job['Status'] == VideoStatus.ERROR:
+            video.update_status(VideoStatus.TRANSCODE_FAILED)
```
Comment (giocalitri) [inline ui/api.py]: do we need to log anything here?

File: ui/models.py [hunk 27]
```diff
@@ -88,6 +88,16 @@ def transcode_key(self, preset=None):
...
 
+    def update_status(self, status):
+        """
+        Assign and save the status of a Video
+
+        Args:
+            status(str): The status to assign the video
+        """
+        self.status = status
+        self.save()
```
Comment (giocalitri) [inline ui/models.py]: we should probably open an issue to modify the status field as a list of options

**Comment by giocalitri** [comment]
There are a couple of very minor stuff, but I see in codecov that `cloudsync/task.py` is missing some tests.

File: ui/models.py (hunk 27 above)
Comment (mbertrand) [inline ui/models.py]: https://github.com/mitodl/odl-video-service/issues/87

File: ui/api.py (hunk 26 above)
Comment (mbertrand) [inline ui/api.py]: Added a log.error statement

**Comment by giocalitri** [comment]
👍 after  squashing
